src/imputation/imputation.py

```diff
@@ -0,0 +1,391 @@
+    for unique_item in df[column].unique():
+        unique_item_df = df[df[column] == unique_item].copy()
+        for target_variable in target_variables_list:
+            growth_ratio_df = calc_growth_ratio(
+                target_variable, unique_item_df, current_quarter, previous_quarter
+            )
+            sorted_df = sort(target_variable, growth_ratio_df)
+            trim_check_df = trim_check(sorted_df)
+            trimmed_df = trim(trim_check_df)
+
+            dict_mean_growth_ratio = get_mean_growth_ratio(
+                trimmed_df, dict_mean_growth_ratio, unique_item, target_variable
+            )
+            # growth_ratio_dfs_list.append(growth_ratio_df)
+            # could also store in a df?
+
+    # growth_ratio_df = pd.concat(growth_ratio_dfs_list)
+    # could also store ina dataframe
+
+    return dict_mean_growth_ratio  # , growth_ratio_df
+    # aka "imputation links" - what naming is best?
+
+
+# TODO break this function into smaller functions
+def forward_imputation(
+    df: pd.DataFrame,
+    column: str,
+    target_variables_list: list,
+    current_quarter: str,
+    previous_quarter: str,
+) -> pd.DataFrame():
+    """_summary_
+
+    Args:
+        df (_type_): _description_
+
+    Returns:
+        _type_: _description_
+    """
+
+    df_growth_ratio = df[
+        df["current_quarter_var1"] != "missing"
+    ].copy()  # TODO add f string
+
+    dict_mean_growth_ratio = loop_unique(
+        df_growth_ratio,
+        column,
+        target_variables_list,
+        current_quarter,
+        previous_quarter,
+    )
+
+    # {'class1_var1_mean_growth_ratio and count':[0.5,11]}
+    dfs_list = []
+    df_final = df.copy()
+    for class_name in df_final[f"{current_quarter}_class"].unique():
+        for var in target_variables_list:
+            print("dmklsd")
+            df_other = df_final[
+                df_final[f"{current_quarter}_class"] == class_name
+            ].copy()
+            df_other = df_other[
+                df_other["current_quarter_var1"] == "missing"
+            ].copy()  # change the name of df_final and df_other
+            df_other[f"{class_name}_{var}_growth_ratio"] = int(
+                dict_mean_growth_ratio[
+                    f"{class_name}_{var}_mean_growth_ratio and count"
+                ][0]
+            )  # why doesn't float work?
+            df_other[f"forwards_imputed_{var}"] = (
+                df_other[f"{class_name}_{var}_growth_ratio"]
+                * df_other[f"{previous_quarter}_{var}"]
+            )
+            df_other = df_other.drop(columns=[f"{class_name}_{var}_growth_ratio"])
+            dfs_list.append(df_other)
+
+    df_out = pd.concat(dfs_list)
+
+    return df_out
+
+
+# TODO break this function into smaller functions
+def backwards_imputation(
+    df: pd.DataFrame,
+    column: str,
+    target_variables_list: list,
+    current_quarter: str,
+    previous_quarter: str,
+) -> pd.DataFrame():
```

Comment:
Again, I think this should be minus the brackets?

Reply:
removed